jctools-experimental/src/main/java/org/jctools/stacks/WFIStack.java

```diff
@@ -0,0 +1,373 @@
+     *
+     * @return An iterable of all the nodes in the stack.
+     */
+    public Iterable<T> popAll()
+    {
+        return new NodeIterable<T>(xchgHead(END));
+    }
+
+    /**
+     * Remove all nodes from the stack, and return an iterable of the removed nodes.
+     * The returned iterable is a complete and atomic snapshot of the stack contents at the time of the method call.
+     * Racing callers will not see any overlap between their returned iterables.
+     * The iterable is re-entrant, so {@link Iterable#iterator()} can be called on it multiple times, and it will
+     * always produce the same result.
+     * <p>
+     * This operation is wait-free and has time-complexity linear to the number of elements in the returned iterable.
+     *
+     * @return An iterable of all the nodes in the stack.
+     */
+    public Iterable<T> popAllFifo()
+    {
+        return new NodeIterable<T>(xchgHead(END)).reverse();
+    }
+
+    /**
+     * Replaces and returns all nodes in the stack with the given node, and return an iterable of the removed nodes.
+     * The returned iterable is a complete and atomic snapshot of the stack contents at the time of the method call.
+     * Racing callers will not see any overlap between their returned iterables.
+     * The iterable is re-entrant, so {@link Iterable#iterator()} can be called on it multiple times, and it will
+     * always produce the same result.
+     * <p>
+     * This operation is wait-free and has constant time-complexity.
+     *
+     * @return An iterable of all the nodes in the stack, in Last In, First Out order.
+     */
+    public Iterable<T> replaceAll(T node)
+    {
+        checkPush(node);
+        node.count = 1;
+        node.next = END;
+        return new NodeIterable<T>(xchgHead(node));
+    }
+
+
+    /**
+     * Replaces and returns all nodes in the stack with the given node, and return an iterable of the removed nodes.
+     * The returned iterable is a complete and atomic snapshot of the stack contents at the time of the method call.
+     * Racing callers will not see any overlap between their returned iterables.
+     * The iterable is re-entrant, so {@link Iterable#iterator()} can be called on it multiple times, and it will
+     * always produce the same result.
+     * <p>
+     * This operation is wait-free and has time-complexity linear to the number of elements in the returned iterable.
+     *
+     * @return An iterable of all the nodes in the stack, in First In, First Out order.
+     */
+    public Iterable<T> replaceAllFifo(T node)
+    {
+        checkPush(node);
+        node.count = 1;
+        node.next = END;
+        return new NodeIterable<T>(xchgHead(node)).reverse();
+    }
+
+    /**
+     * Pop, or remove, the most recently pushed node, or return {@code null} if the stack is empty.
+     * <p>
+     * This operation is non-blocking, but not wait-free. It has constant time-complexity.
```

Comment:
I would say it's lock-free (in the "WF > LF > blocking" sense). I'm also not sure the "constant time complexity" is informative or correct here. The worst case time here is infinite through no relation to N IIUC.